### PR TITLE
planner: add privilege check for lock/unlock stats

### PR DESCRIPTION
**Pull request by hi-rustin**
### What problem does this PR solve?

Issue Number: ref https://github.com/pingcap/tidb/issues/46351

Problem Summary:

### What is changed and how it works?

Added privilege check for lock/unlock stats. They require the `SELECT` and `INSERT` privileges.

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [x] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Comment by hi-rustin** [comment]
> 1. Start the TiDB cluster: `tiup playground nightly --tiflash 0 --db.binpath /Users/hi-rustin/GolandProjects/tidb/bin/tidb-server`
> 2. Create table and analyze it:
> 
> ```sql
> create table t(a int, b int);
> insert into t values (1,2), (3,4), (5,6), (7,8);
> analyze table t;
> show warnings;
> ```
> 
> 3. Lock table: `lock stats t;`
> 4. Analyze it again: `analyze table t;`
> 5. Check the warnings: `show warnings;`
> 
> ```sql
> mysql> show warnings;
> +---------+------+-----------------------------------------------------------------------------------------------------------------------------------------+
> | Level   | Code | Message                                                                                                                                 |
> +---------+------+-----------------------------------------------------------------------------------------------------------------------------------------+
> | Note    | 1105 | Analyze use auto adjusted sample rate 1.000000 for table test.t, reason to use this rate is "use min(1, 110000/4) as the sample-rate=1" |
> | Warning | 1105 | skip analyze locked table: t                                                                                                            |
> +---------+------+-----------------------------------------------------------------------------------------------------------------------------------------+
> ```
> 
> 6. Check the `mysql.stats_table_locked` table:
> 
> ```sql
> mysql> use mysql;
> Reading table information for completion of table and column names
> You can turn off this feature to get a quicker startup with -A
> 
> Database changed
> mysql> select * from stats_table_locked;
> +----------+--------------+-------+--------------------+
> | table_id | modify_count | count | version            |
> +----------+--------------+-------+--------------------+
> |       98 |            4 |     4 | 443863948666601483 |
> +----------+--------------+-------+--------------------+
> 1 row in set (0.00 sec)
> ```
> 
> 7. Try to lock it again: `lock stats t;`
> 
> ```sql
> mysql> use test;
> Reading table information for completion of table and column names
> You can turn off this feature to get a quicker startup with -A
> 
> Database changed
> mysql> lock stats t;
> Query OK, 0 rows affected, 1 warning (0.01 sec)
> 
> mysql> show warnings;
> +---------+------+-----------------------------------+
> | Level   | Code | Message                           |
> +---------+------+-----------------------------------+
> | Warning | 1105 | skip locking locked table: test.t |
> +---------+------+-----------------------------------+
> 1 row in set (0.00 sec)
> ```
> 
> 8. Show locked stats: `show stats_locked;`.
> 
> ```sql
> mysql> show stats_locked;
> +---------+------------+----------------+--------+
> | Db_name | Table_name | Partition_name | Status |
> +---------+------------+----------------+--------+
> | test    | t          |                | locked |
> +---------+------------+----------------+--------+
> 1 row in set (0.00 sec)
> ```
> 
> 9. Unlock the table: `unlock stats t;`
> 10. Check the `mysql.stats_table_locked` table again:
> 
> ```sql
> mysql> select * from stats_table_locked;
> Empty set (0.00 sec)
> ```
> 
> 11. Show locked stats: `show stats_locked;`
> 
> ```sql
> mysql> show stats_locked;
> Empty set (0.00 sec)
> ```
> 
> 12. Unlock the table again and check warnings: `unlock stats t;`
> 
> ```sql
> mysql> unlock stats t;
> Query OK, 0 rows affected, 1 warning (0.00 sec)
> 
> mysql> show warnings;
> +---------+------+---------------------------------------+
> | Level   | Code | Message                               |
> +---------+------+---------------------------------------+
> | Warning | 1105 | skip unlocking unlocked table: test.t |
> +---------+------+---------------------------------------+
> 1 row in set (0.00 sec)
> ```

**Review by hi-rustin — COMMENTED** [review]
🔢 self-check

File: planner/core/planbuilder.go [hunk 1]
```diff
@@ -4530,16 +4521,35 @@ func (*PlanBuilder) buildLoadStats(ld *ast.LoadStatsStmt) Plan {
...
 	return p
 }
 
+// requireInsertAndSelectPriv requires INSERT and SELECT privilege for the tables.
+func (b *PlanBuilder) requireInsertAndSelectPriv(tables []*ast.TableName) {
+	for _, tbl := range tables {
+		user := b.ctx.GetSessionVars().User
+		var insertErr, selectErr error
+		if user != nil {
+			insertErr = ErrTableaccessDenied.GenWithStackByArgs("INSERT", user.AuthUsername, user.AuthHostname, tbl.Name.O)
```
[the remark this answers is not in the document]
Comment (qw4990) [reply planner/core/planbuilder.go:4545]: It seems OK to keep it unchanged.

**Comment by qw4990** [comment]
> 1. Start the TiDB cluster: `tiup playground nightly --tiflash 0 --db.binpath /Users/hi-rustin/GolandProjects/tidb/bin/tidb-server`
> 2. Create table and analyze it:
> 
> ```sql
> create table t(a int, b int);
> insert into t values (1,2), (3,4), (5,6), (7,8);
> analyze table t;
> show warnings;
> ```
> 
> 3. Create a user: `CREATE USER 'dd'@'localhost' IDENTIFIED BY 'password';`
> 4. Only give it delete privilege: `GRANT DELETE ON test.* TO 'dd'@'localhost'; FLUSH PRIVILEGES;`
> 5. Try to lock or unlock the table t: `lock stats t;`/ `unlock stats t;`
> 
> ```sql
> mysql> lock stats t;
> ERROR 1142 (42000): INSERT command denied to user 'dd'@'localhost' for table 't'
> mysql> unlock stats t;
> ERROR 1142 (42000): INSERT command denied to user 'dd'@'localhost' for table 't'
> ```
> 
> 6. Give it the `INSERT` privilege: `GRANT INSERT ON test.* TO 'dd'@'localhost'; FLUSH PRIVILEGES;`
> 7. Try to lock or unlock the table t: `lock stats t;`/ `unlock stats t;`
> 
> ```sql
> mysql> lock stats t;
> ERROR 1142 (42000): SELECT command denied to user 'dd'@'localhost' for table 't'
> mysql> unlock stats t;
> ERROR 1142 (42000): SELECT command denied to user 'dd'@'localhost' for table 't'
> ```
> 
> 8. Give it the `SELECT` privilege: `GRANT SELECT ON test.* TO 'dd'@'localhost'; FLUSH PRIVILEGES;`
> 9. Lock table: `lock stats t;`
> 10. Analyze it again: `analyze table t;`
> 11. Check the warnings: `show warnings;`
> 
> ```sql
> mysql> show warnings;
> +---------+------+-----------------------------------------------------------------------------------------------------------------------------------------+
> | Level   | Code | Message                                                                                                                                 |
> +---------+------+-----------------------------------------------------------------------------------------------------------------------------------------+
> | Note    | 1105 | Analyze use auto adjusted sample rate 1.000000 for table test.t, reason to use this rate is "use min(1, 110000/4) as the sample-rate=1" |
> | Warning | 1105 | skip analyze locked table: t                                                                                                            |
> +---------+------+-----------------------------------------------------------------------------------------------------------------------------------------+
> ```
> 
> 12. Try to lock it again: `lock stats t;`
> 
> ```sql
> mysql> use test;
> Reading table information for completion of table and column names
> You can turn off this feature to get a quicker startup with -A
> 
> Database changed
> mysql> lock stats t;
> Query OK, 0 rows affected, 1 warning (0.01 sec)
> 
> mysql> show warnings;
> +---------+------+-----------------------------------+
> | Level   | Code | Message                           |
> +---------+------+-----------------------------------+
> | Warning | 1105 | skip locking locked table: test.t |
> +---------+------+-----------------------------------+
> 1 row in set (0.00 sec)
> ```
> 
> 13. Show locked stats: `show stats_locked;`.
> 
> ```sql
> mysql> show stats_locked;
> ERROR 1142 (42000): SHOW command denied to user 'dd'@'localhost' for table 'stats_table_locked'
> ```
> 
> 14. Give it the `SELECT` privilege on `mysql.stats_table_locked`: `GRANT SELECT ON mysql.stats_table_locked TO 'dd'@'localhost'; FLUSH PRIVILEGES;`
> 15. Show locked stats: `show stats_locked;`:
> 
> ```sql
> mysql> show stats_locked;
> +---------+------------+----------------+--------+
> | Db_name | Table_name | Partition_name | Status |
> +---------+------------+----------------+--------+
> | test    | t          |                | locked |
> +---------+------------+----------------+--------+
> 1 row in set (0.00 sec)
> ```
> 
> 16. Unlock the table: `unlock stats t;`
> 17. Check the `mysql.stats_table_locked` table again:
> 
> ```sql
> mysql> select * from mysql.stats_table_locked;
> Empty set (0.00 sec)
> ```
> 
> 18. Show locked stats: `show stats_locked;`
> 
> ```sql
> mysql> show stats_locked;
> Empty set (0.00 sec)
> ```
> 
> 19. Unlock the table again and check warnings: `unlock stats t;`
> 
> ```sql
> mysql> unlock stats t;
> Query OK, 0 rows affected, 1 warning (0.00 sec)
> 
> mysql> show warnings;
> +---------+------+---------------------------------------+
> | Level   | Code | Message                               |
> +---------+------+---------------------------------------+
> | Warning | 1105 | skip unlocking unlocked table: test.t |
> +---------+------+---------------------------------------+
> 1 row in set (0.00 sec)
> ```

LGTM, but could you transform this to a unit test like `TestTableLockPrivilege`?

File: planner/core/planbuilder.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (hi-rustin) [reply planner/core/planbuilder.go:4545]: Got it. Thanks!

**Comment by hi-rustin** [comment]
> LGTM, but could you transform this to a unit test like `TestTableLockPrivilege`?

Good point. I will add it.

**Review by qw4990 — APPROVED** [review]

**Review by hawkingrei — APPROVED** [review]

**Pull request merged**